### ISSUE TITLE
Question about implementation

### ISSUE THREAD

**Issue by hychaochao** [FYYDCC/IVT-LR#1]
Very nice work with a lot of valuable insights! I have a couple of questions about the implementation:

1. In the forward loop, `end` is updated as: `end = end + 1 + K`
https://github.com/FYYDCC/IVT-LR/blob/a47db8c6c8d03584b040617520f84cf3e4234a7a/qwen_vl/qwen_ivtlr.py#L264-L267 
However, I'm wondering if `end` should instead be set to the position of the next latent token, something like:
`end = latent_lists[b][pass_idx + 1] `

2. I noticed that when substituting the latent token with the hidden state of the previous token, the original tokens before the latent token remain in the sequence instead of being removed:
https://github.com/FYYDCC/IVT-LR/blob/a47db8c6c8d03584b040617520f84cf3e4234a7a/qwen_vl/qwen_ivtlr.py#L269-L289
This means that when computing the final logits after all loops finish, all tokens are attended to, rather than only retaining the latent tokens that contain the compressed information (as done in Coconut). I wonder why?

I may have misunderstood the implementation, and I'd appreciate any clarification you can provide. Thank you!

**Comment by FYYDCC** [comment]
Thank you very much for the careful reading and for raising these thoughtful questions

1. We already know:
- the number of latent tokens (max_n_latents)
- the number of visual embeddings per latent token (K)

Each iteration of the loop is designed to:

- consume one latent token
- plus its associated K visual embeddings

Therefore, the sequence growth pattern is fixed and deterministic:
each iteration extends the effective sequence length by 1 + K,
end = end + 1 + K is sufficient and avoids extra indexing into latent_lists

The reason for not using the latent token positions is implementation convenience. If one chooses to use them, it is also feasible — after each selection, the positions of the remaining latent tokens just need to be updated by adding a +K offset.


2. Regarding the second question, this behavior is intentional and differs from Coconut. In our implementation, we do not rely on KV-cache, and each forward pass recomputes attention from scratch. The visual embeddings are re-selected and re-inserted at every iteration rather than being stored in the cache. As a result, the original tokens need to remain in the sequence; otherwise, their information would be lost. In contrast, Coconut uses KV-cache to preserve past information, which allows non-latent tokens to be dropped. This is why, after all loops finish, the model still attends to all tokens rather than only the latent tokens.
